### PR TITLE
Make it possible to wipe all tokens/devices of a user as (sub) admin

### PR DESCRIPTION
**Pull request by ChristophWurst**
This PR consists of two parts. The first commit refactors the *mark token for remote wipe* logic into the remote wipe service. The second commit adds the new feature.

### CODE REVIEW

**Review by skjnldsv — APPROVED** [review]
Code looks spot on! :+1:

File: apps/provisioning_api/appinfo/routes.php [hunk 1]
```diff
@@ -50,6 +50,7 @@
 		['root' => '/cloud', 'name' => 'Users#getCurrentUser', 'url' => '/user', 'verb' => 'GET'],
 		['root' => '/cloud', 'name' => 'Users#getEditableFields', 'url' => '/user/fields', 'verb' => 'GET'],
 		['root' => '/cloud', 'name' => 'Users#editUser', 'url' => '/users/{userId}', 'verb' => 'PUT'],
+		['root' => '/cloud', 'name' => 'Users#wipeUserDevices', 'url' => '/users/{userId}/wipe', 'verb' => 'POST'],
```
Comment (csware) [inline apps/provisioning_api/appinfo/routes.php]: The name could indicate that it is possible to delete the devices themselves.

**Review by rullzer — APPROVED** [review]
So it works as advertised. But the UX is a bit weird. AFter you press the 'wipe devices' button it spins quick and then reloads. No feedback if something had been done etc.

**Comment by ChristophWurst** [comment]
> No feedback if something had been done etc.

Right. Any suggestions? A simple notification?

**Comment by rullzer** [comment]
> > No feedback if something had been done etc.
> 
> Right. Any suggestions? A simple notification?

probably a good step yes. Then we can go from there.

File: lib/private/Authentication/Token/RemoteWipe.php [hunk 2]
```diff
@@ -25,19 +25,15 @@
 
 namespace OC\Authentication\Token;
 
-use BadMethodCallException;
+use function array_filter;
```
Comment (MorrisJobke) [inline lib/private/Authentication/Token/RemoteWipe.php]: haven't seen that before, but isn't an inline `\array_filter` easier to read and maintain?

**Review by MorrisJobke — APPROVED** [review]
Code looks good 👍

File: .htaccess [hunk 3]
```diff
@@ -65,3 +65,7 @@ Options -Indexes
 <IfModule pagespeed_module>
   ModPagespeed Off
 </IfModule>
+#### DO NOT CHANGE ANYTHING ABOVE THIS LINE ####
+
+ErrorDocument 403 /
+ErrorDocument 404 /
```
Comment (MorrisJobke) [inline .htaccess]: 🙈

Comment (rullzer) [reply .htaccess]: A glitch in the matrix...

**Pull request merged**